src/datasources/asset/AssetDataSource.ts

```diff
@@ -0,0 +1,111 @@
+  }
+
+  async processMetadataQuery(query: AssetQuery) {
+    const result: DataFrameDTO = { refId: query.refId, fields: [] };
+    let workspaceId = this.templateSrv.replace(query.workspace);
+    const minionIds = replaceVariables(query.minionIds, this.templateSrv);
+    const conditions = [];
+    if (workspaceId) {
+      conditions.push(`workspace = "${workspaceId}"`);
+    }
+    if (minionIds.length) {
+      const systemsCondition = minionIds.map(id => `${AssetFilterProperties.LocationMinionId} = "${id}"`)
+      conditions.push(`(${systemsCondition.join(' or ')})`);
+    }
+    const assetFilter = conditions.join(' and ');
+    const assets: AssetModel[] = await this.queryAssets(assetFilter);
+    const workspaces = await this.getWorkspaces();
+    result.fields = [
+      { name: 'id', values: assets.map(a => a.id) },
+      { name: 'name', values: assets.map(a => a.name) },
+      { name: 'model name', values: assets.map(a => a.modelName) },
+      { name: 'serial number', values: assets.map(a => a.serialNumber) },
+      { name: 'bus type', values: assets.map(a => a.busType) },
+      { name: 'asset type', values: assets.map(a => a.assetType) },
+      { name: 'is NI asset', values: assets.map(a => a.isNIAsset) },
+      { name: 'calibration status', values: assets.map(a => a.calibrationStatus) },
+      { name: 'is system controller', values: assets.map(a => a.isSystemController) },
+      { name: 'last updated timestamp', values: assets.map(a => a.lastUpdatedTimestamp) },
+      { name: 'minionId', values: assets.map(a => a.location.minionId) },
+      { name: 'parent name', values: assets.map(a => a.location.parent) },
+      { name: 'system name', values: assets.map(a => a.location.systemName) },
+      {
+        name: 'calibration due date',
+        values: assets.map(a => a.externalCalibration ? a.externalCalibration.nextRecommendedDate : null)
+      },
+      { name: 'workspace', values: assets.map(a => getWorkspaceName(workspaces, a.workspace)) },
+    ];
+    return result;
+  }
+
+
+  shouldRunQuery(_: AssetQuery): boolean {
+    return true;
+  }
+
+  async queryAssets(filter = '', take = -1): Promise<AssetModel[]> {
+    let data = { filter, take };
+    try {
+      let response = await this.post<AssetsResponse>(this.baseUrl + '/query-assets', data);
+      return response.assets;
+    } catch (error) {
+      throw new Error(`An error occurred while querying assets: ${error}`);
+    }
+  }
+
+  async querySystems(filter = '', projection = defaultProjection): Promise<SystemMetadata[]> {
+    try {
+      let response = await this.getSystems({
+        filter: filter,
+        projection: `new(${projection.join()})`,
+        orderBy: defaultOrderBy,
+      })
+
+      return response.data;
+    } catch (error) {
+      throw new Error(`An error occurred while querying systems: ${error}`);
+    }
+  }
+
+  async testDatasource(): Promise<TestDataSourceResponse> {
+    await this.get(this.baseUrl + '/assets?take=1');
```

Comment:
Is this GET going to throw an error in case of 4xx / 5xx status codes ?

Shouldn't we at least check that the response is a valid result before returning success ?

Reply:
I'm sure the testDatasource function is only runs when we check the validity of the credentials, it throws a 401 when the user is not authorized, same as rest of testDatasource in our other plugins